商品購入機能の実装

shiki0501 opened the pull request:
# what
商品購入機能実装
# why
商品購入機能の実装をし、ユーザー間での売買を成立させるため

- 正しく情報を記入すると、商品の購入ができる動画
https://gyazo.com/dee7872d330055df233b57c991b55989
- 入力に問題のある状態では商品の購入はできず、エラーメッセージが出力される動画
https://gyazo.com/7c728e7b86d945ef142d23d0a8822d83
- ログアウト状態のユーザーが、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/6a92744be4951dadcd348754949e562d
- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
https://gyazo.com/107831f32f31132fa1f8deeac136a008
- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/cb292c51d5c548544bb7ac777432ba2f
- 売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/f0422b94e8f747d1b170c55a9e875133
- 売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/a4f6367c892726c39927b01f6e0b1065
- ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/6bc1b9e84fef50c53f26b4e727233cd4
- ログイン状態の出品者以外のユーザーでも、売却済みの商品に対しては「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/087492e9e1205500b9465e31215af8d8
- ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/128b34bc8233e5128873e89000a80be1
- テスト結果の画像
https://gyazo.com/32ee9c5789521e75c2de4d843d0abf86

コードレビューのほどよろしくお願いいたします。

shiki0501 committed:
delivery,buyモデル作成

モデル作成、データベース追加
rubocop実行

shiki0501 committed:
deliveryコントローラ作成

deliveryコントローラ作成
・indexファイル追加
・routesの追加

shiki0501 committed:
購入画面への遷移、アソシエーションの追加、購入画面の画像等表示

購入画面への遷移
アソシエーションの追加
・各modelへアソシエーション記述
購入画面の表示
・routesにネストを組むことで表現
ネスト
https://qiita.com/chopesu_se/items/c7362380865cf978b158

shiki0501 committed:
ActiveHashの導入

shiki0501 committed:
API導入

javascriptの記述修正
・手作業でJavaScriptを記述してフォーム送信処理などを実装する場合は、このturbolinksがその処理を無効にしてしまうことがあるため、turbolinksの記述を削除
card.jsファイルの追加

shiki0501 committed:
API導入２

・10〜15
フォーム画面から入力させた情報をcardへ入れる。
・18〜23
トークン情報をフォームに追加
・26〜２９
クレジット情報の削除
入力された各情報を削除（セキュリティ性アップ）

shiki0501 committed:
API導入３

・createアクション追加、route設定
・index.erbへ各カラム名称追加
・クレジットカード　削除記述修正
※formオブジェクト実装を先にやらなければ、paramsへデータが遅れないため、中断。。

shiki0501 committed:
APIキー変更

間違えてコミットしたので、変更。

shiki0501 committed:
formオブジェクト

・formオブジェクト作成
・formオブジェクトを使用したアクション、erbファイル修正、params設定（バリテーションはまだ）
・データベースへの登録ができないので、その機能は要確認！！

shiki0501 committed:
バリテーションの追加

・空白では登録出来ない
・エラーログが画面に表示される
・token（クレカ情報）をdelivery modelに追加

shiki0501 committed:
クレジットカード購入機能実装

・payjp導入
・card.jp記述変更
　name属性変更（deleivery_buy)
・controller記述追加
　Payjpへ売り上げが反映される。

※現状はエラー発生し、データベースへは保存されない。

shiki0501 committed:
商品購入機能実装

・user_id,item_id　merge
・deliverybuy model修正
・各情報がDBへ保存されることを確認

shiki0501 committed:
商品一覧機能

- 売却済みの商品は、画像上に『sold out』の文字が表示されるようになっていること（商品一覧機能）

shiki0501 committed:
商品詳細機能の実装

- 売却済みの商品は、画像上に『sold out』の文字が表示されるようになっていること（商品詳細機能）

shiki0501 committed:
商品一覧・詳細画面　修正

soldout表示　修正
商品編集・削除表示　修正

shiki0501 committed:
画面遷移、ボタン表示

・売却済みは直接URLを入れても編集画面遷移しない
・売却済み商品は購入画面に進むボタンが表示されない

shiki0501 committed:
単体テストコードの記述

・バリテーションの修正
・単体テストコードの記述

shiki0501 committed:
環境設定の導入

クレカのトークン情報の環境設定化

shiki0501 committed:
アクセス制限

- ログアウト状態のユーザーは、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移すること
- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移すること

shiki0501 committed:
rubocop使用

 bundle exec rubocop -a　にてまとめ

shiki0501 committed:
サーバー、クライアントサーバでの環境変数呼び出し

各記述にて環境変数の読み出しをする

shiki0501 committed:
コメントアウトの修正、アクセス制限

・card.js確認のためコメントアウトしていたので戻した
・アクセス制限の記述
redirect_to root_path

shiki0501 committed:
アクセス制限修正、rubocop

shiki0501 committed:
不要なコメントアウトの削除

tomi0420 commented on app/controllers/deliverys_controller.rb:
```diff
@@ -0,0 +1,35 @@
+  before_action :authenticate_user!, only: [:index]
+
+  def index
+    @item = Item.find(params[:item_id])
```
```
@item = Item.find(params[:item_id])
```

こちらですが、createアクションにも同様の記述があるため、before_actionでまとめましょう。
理由：
同じ記述をまとめる事で、可読性が上がるためです。

tomi0420 commented on app/controllers/deliverys_controller.rb:
```diff
@@ -0,0 +1,35 @@
+      @delivery_buy.save
+      redirect_to root_path
+    else
+      @item = Item.find(params[:item_id])
```
```
@item = Item.find(params[:item_id])
```
こちらの記述ですが、削除していただいて構いません。
理由：
createアクション内（もしくはbefore_action内）で@itemが定義されていれば、renderでビューを呼び出す前に、再度定義する必要がないためです。

tomi0420 commented on app/models/delivery_buy.rb:
```diff
@@ -0,0 +1,20 @@
+    validates :municipality
+    validates :address
+    validates :tell, format: { with: /\A\d{11}\z/, message: 'is invalid.Only numbers within 11 digits can be saved' }
+  end
```
```
・user_id
・item_id
```

に関しても、「presence: true」のバリデーションをかけましょう。
理由：
現状の記述ですと、「ユーザー情報」と「商品情報」がなくても購入できてしまうためです。

tomi0420 commented on app/models/delivery_buy.rb:
```diff
@@ -0,0 +1,20 @@
+    validates :shipping_area_id, numericality: { other_than: 1, message: "can't be blank" }
+    validates :municipality
+    validates :address
+    validates :tell, format: { with: /\A\d{11}\z/, message: 'is invalid.Only numbers within 11 digits can be saved' }
```
```
validates :tell, format: { with: /\A\d{11}\z/, message: 'is invalid.Only numbers within 11 digits can be saved' }
```

電話番号に関して、
```
10桁以上11桁以下であること
```
になるように修正しましょう。

tomi0420 commented on app/views/items/show.html.erb:
```diff
@@ -25,13 +25,15 @@
+          <%= link_to "購入画面に進む", item_deliverys_path(@item) ,class:"item-red-btn"%>
+        <% end %>
       <% end %>
     <% end %>
```
こちらの条件分岐ですが、以下のような構造でリファクタリングいただければと思います。
```
if "ログインしていれば" かつ "商品は売れていなければ" 
  if "現在ログインしているユーザーが商品の出品者であれば"
    商品の編集
    削除
  else
    購入画面に進む
  end
end
```
１．大きな条件として「"ログインしていれば" かつ "商品は売れていなければ" 」
　こちらの条件で「商品の編集」「削除」「購入画面に進む」が表示されるかを制御しましょう。
２．１の条件の中でさらに条件を追加します。
　「"現在ログインしているユーザーが商品の出品者であれば"」
　こちらの条件で「商品の編集」「削除」を表示させるか
　 「購入画面に進む」を表示させるかを制御しましょう。

tomi0420 commented on spec/models/delivery_buy_spec.rb:
```diff
@@ -0,0 +1,73 @@
+    before do
+      user = FactoryBot.create(:user)
+      @delivery_buy = FactoryBot.build(:delivery_buy)
+    end
```
```
before do
  user = FactoryBot.create(:user)
  @delivery_buy = FactoryBot.build(:delivery_buy)
end
```

今回購入において商品やユーザーを用意する必要があるのであれば、以下を参考にdelivery_buy__spec.rb のbefore内でユーザーや商品のFactoryBotから生成しましょう。
```
1. ユーザーのダミーデータを作成し、変数に格納する
2. アイテムのダミーデータを作成し、変数に格納する
3. @△△ = FactoryBot.build(:〇〇)にuser_idとitem_idをキーとし、それぞれのバリューとして1,2で定義した変数からidを抜き出す記述をしましょう。
4. エラー回避のために、sleepメソッドを使用しましょう。
```
ご不明な点がありましたら、コネクトにてメンターにご相談ください！

tomi0420 commented on spec/models/delivery_buy_spec.rb:
```diff
@@ -0,0 +1,73 @@
+      end
+    end
+
+    context '内容に問題がある場合' do
```
異常系のテストに関して、以下の内容を追加しましょう！
```
・電話番号が9桁以下では購入できない
・電話番号が12桁以上では購入できない
・userが紐づいていなければ購入できない
・itemが紐づいていなければ購入できない
```

tomi0420 reviewed (commented):
レビュー依頼ありがとうございます！
コメントいたしましたので、ご確認をお願い致します。

shiki0501 committed:
コードレビューの修正

delivery_controller
・before actionによる可読性アップ
delivery_buy model
・item_id,user_id　バリテーション追加（空白では登録出来ない）
・tell 10桁以上１１桁未満でないと登録出来ない
show.html.erb
・「"ログインしていれば" かつ "商品は売れていなければ" 」「商品の編集」「削除」「購入画面に進む」が表示される
・「"現在ログインしているユーザーが商品の出品者であれば"」「商品の編集」「削除」を表示させるか購入画面に進む」を表示させる
spec
・各単体テストコード追加
user_id,item_idの紐つけ、テストコード追加

shiki0501 committed:
rubocop使用

shiki0501 commented:
コードレビューありがとうございました！
修正いたしましたので、再レビューのほどよろしくお願いします。

yuri240 commented on app/models/delivery_buy.rb:
```diff
@@ -0,0 +1,22 @@
+class DeliveryBuy
+  include ActiveModel::Model
+  attr_accessor :postal_code, :shipping_area_id, :municipality, :address, :building_name, :tell, :buy_id, :user_id, :item_id,
```
以下は、**任意修正**となります。

`:buy_id,`

attr_accessorに記載されている`:buy_id`は不要と思われますので
削除していただけると良いですね！

理由：
`:buy_id`はparamsのデータとして送られてこない情報であるため。

yuri240 commented on spec/models/delivery_buy_spec.rb:
```diff
@@ -0,0 +1,85 @@
+        expect(@delivery_buy.errors.full_messages).to include("Shipping area can't be blank")
+      end
+      it 'shipping_area_idが1だと保存できないこと' do
+        @delivery_buy.shipping_area_id = '1'
```
以下は、**任意修正**となります。


```ruby
@delivery_buy.shipping_area_id = '1'
```
アクティブハッシュの値は、integer型で数値として扱うため
クォーテーション(`'`)は外してあげましょう！

理由：
`'`がついていることで文字列として扱われてしまうため。

yuri240 reviewed (commented):
修正とレビュー依頼ありがとうございます。

しっかりと実装できていることが確認できましたので、
**LGTM**とさせていただきます！

また、任意修正箇所として、
２点コメントを残させていただきましたので
確認の上でMergeしていただくようお願いいたします。

ここまでの実装、大変お疲れ様でした。
引き続き、応援しております。

shiki0501 committed:
コードレビュー修正

・buy_id削除　
buy_idはparamsのデータとして送られてこない情報であるため。
・''削除
'がついていることで文字列として扱われてしまうため。
